sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/SaplingParamTool.kt

```diff
@@ -31,30 +30,163 @@ object SaplingParamTool {
      */
     internal const val OUTPUT_PARAM_FILE_MAX_BYTES_SIZE = 5L * 1024L * 1024L
 
+    /**
+     * Subdirectory name, in which are the sapling params files stored.
+     */
+    internal const val SAPLING_PARAMS_LEGACY_SUBDIRECTORY = "params"
+
+    /**
+     * File name for the sapling spend params
+     */
+    internal const val SPEND_PARAM_FILE_NAME = "sapling-spend.params"
+
+    /**
+     * File name for the sapling output params
+     */
+    internal const val OUTPUT_PARAM_FILE_NAME = "sapling-output.params"
+
+    /**
+     *  Temporary file prefix to fulfill atomicity requirement of file handling
+     */
+    private const val TEMPORARY_FILE_NAME_PREFIX = "_"
+
+    /**
+     * File SHA1 hash for the sapling spend params
+     */
+    internal const val SPEND_PARAM_FILE_SHA1_HASH = "a15ab54c2888880e53c823a3063820c728444126"
+
+    /**
+     * File SHA1 hash for the sapling output params
+     */
+    internal const val OUTPUT_PARAM_FILE_SHA1_HASH = "0ebc5a1ef3653948e1c46cf7a16071eac4b7e352"
+
+    /**
+     * The Url that is used by default in zcashd
+     */
+    private const val CLOUD_PARAM_DIR_URL = "https://z.cash/downloads/"
+
+    private val checkFilesMutex = Mutex()
+
+    internal var toolsProperties: SaplingParamToolProperties? = null
+
+    /**
+     * Initialization of needed properties. This is necessary entry point for other operations from {@code
+     * SaplingParamTool}. This type of implementation also simplifies its testing.
+     *
+     * @param context
+     */
+    internal suspend fun initSaplingParamTool(context: Context): SaplingParamToolProperties {
```

Comment:
Do you think we could eliminate the global state of `var toolsProperties`?

I'm wondering whether something like:
 - Make `SaplingParamTool` a class
 - Create a private constructor with `SaplingParamToolProperties`
 - Move this method and the one below to a companion object which acts as a factory for returning a SaplingParamTool instance

Reply:
Yea, I was thinking about making this whole SaplingParamTool class and calling its methods better too. Here are some of the implementation requirements, which led to what we have now:
- instantiate this class from a context class (i.e. `Initializer.kt`)
- do context operation (i.e. create needed directories: cache/no_backup)
- save result of these context operations (i.e. `toolProperties`)
- then call other methods (i.e. `ensureParams` and `fetchParams`) from non-context classes (i.e. `WalletTransactionEncoder.kt`), which uses the saved result of context operations
- we need to access the `toolProperties` also from tests

We probably can't move the `toolProperties` to class, as it wouldn't be accessible from static methods. If you can see any improvement to this, I'll be glad. Thank you.